### PR TITLE
Refrigerator Malfunction Workflow

### PR DESCRIPTION
**Pull request by gcapalbo**
@czue 

Don't merge yet - just opening for review. This includes the entire SMS workflow from the facility user's perspective. What's left is to do the district user's side of the SMS workflow and add some tests.

### CODE REVIEW

**Comment by czue** [comment]
made (and added) a "don't merge" label we can use

File: logistics_project/apps/malawi/handlers/abstract/base.py [hunk 1]
```diff
@@ -14,4 +15,24 @@ def respond(self, template, **kwargs):
...
+    Intended to be used when Facility users need to interact with the system
+    over SMS.
+    """
+
+    def validate_contact(self):
+        """
+        Checks if the contact who sent this message can use the keyword
+        and returns that contact's supply point if so.
+        """
+        contact = getattr(self.msg,'logistics_contact', None)
```
Comment (czue) [inline logistics_project/apps/malawi/handlers/abstract/base.py]: you can add `@logistics_contact_required` to the `handle` function ensure it is always set (and respond if not)

File: logistics_project/apps/malawi/handlers/refrigerator_malfunction.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        return False
+
+    def handle(self, text):
+        supply_point = self.validate_contact()
+        if supply_point:
+            words = text.split()
+            if self.is_text_valid(words):
+                reason = words[0]
+                if not self.malfunction_exists(supply_point):
```
Comment (czue) [inline logistics_project/apps/malawi/handlers/refrigerator_malfunction.py]: the else of this should probably do something... (either respond saying it's already reported, change the reason, or something else - maybe a spec clarification?)

File: logistics_project/apps/malawi/models.py [hunk 3]
```diff
@@ -13,5 +14,67 @@ class Organization(models.Model):
...
+    REASON_FRIDGE_BREAKDOWN = '3'
+
+    REASONS = (
+        REASON_NO_GAS,
+        REASON_POWER_FAILURE,
+        REASON_FRIDGE_BREAKDOWN,
+    )
+
+    # A reference to the facility with the malfunction
+    supply_point = models.ForeignKey('logistics.SupplyPoint', db_index=True, related_name='+')
```
Comment (czue) [inline logistics_project/apps/malawi/models.py]: never knew about `related_name='+'` before.

File: logistics_project/apps/malawi/models.py [hunk 4]
```diff
@@ -13,5 +14,67 @@ class Organization(models.Model):
...
+
+    # Timestamp when the facility user reported the malfunction
+    reported_on = models.DateTimeField(db_index=True)
+
+    # Takes the value of one of the REASON_* constants above
+    malfunction_reason = models.CharField(max_length=1)
```
Comment (czue) [inline logistics_project/apps/malawi/models.py]: you can use `choices=` to get some built in validation and UI stuff for this.

File: static/malawi/config.py [hunk 5]
```diff
@@ -259,11 +259,44 @@ class Messages(object):
...
                 "'help start' or 'help stop' to start and stop reminders."
 
+    # "rm" keyword
+    # Responses to Facility user reporting the refrigerator malfunction
+    FRIDGE_FACILITY_NO_GAS = "Thank you. The district will be informed that your refrigerator is out of gas."
+    FRIDGE_FACILITY_POWER_FAILURE = "Thank you. The district will be informed that your refrigerator has a power failure."
+    FRIDGE_FACILITY_BREAKDOWN = "Thank you. The district will be informed that your refrigerator is not working."
+
+    # Message to the District user notifying of the refrigerator malfunction
+    FRIDGE_DISTRICT_NO_GAS = "%(facility)s has refrigerator malfunction due to no gas."
```
Comment (czue) [inline static/malawi/config.py]: like I mentioned in the gdoc, we should probably include the facility code here. We might want to say something like like "respond with transfer %(code)s [to facility]" at the end of the message.

File: static/malawi/config.py [hunk 6]
```diff
@@ -259,11 +259,44 @@ class Messages(object):
...
+    FRIDGE_DISTRICT_POWER_FAILURE = "%(facility)s has refrigerator malfunction due to power failure."
+    FRIDGE_DISTRICT_BREAKDOWN = "%(facility)s has refrigerator malfunction due to breakdown."
```
Comment (czue) [inline static/malawi/config.py]: not blocking, but we might be able to just define one message and template out the reason which would be a little easier to maintain if there were more reasons later.

File: static/malawi/config.py [hunk 7]
```diff
@@ -259,11 +259,44 @@ class Messages(object):
...
+
+    FRIDGE_HELP = "To report a refrigerator malfunction, please send 'rm [reason code]'"
```
Comment (czue) [inline static/malawi/config.py]: (if there's room we could include the codes here)

File: static/malawi/config.py [hunk 8]
```diff
@@ -259,11 +259,44 @@ class Messages(object):
...
+        "products from %(facility)s")
+
+    FRIDGE_CONFIRM_PRODUCTS_COLLECTED = "Please reply with 'rc' to confirm you have collected your EPI products"
+
+    FRIDGE_NOT_REPORTED_FIXED = ("Your refrigerator has not been reported as fixed yet. If it has been "
+        "fixed, please respond with 'rf'")
+
+    FRIDGE_ALREADY_CONFIRMED_COLLECTED = "You have already confirmed that you have collected your EPI products."
+
+    FRIDGE_CONFIRMATION_RESPONSE = "Thank you for confirming that you have collected your EPI products."
```
Comment (czue) [inline static/malawi/config.py]: at some point we should just get input on the exact words of all these messages but no need to change anything for now.

**Comment by czue** [comment]
looking good!

File: logistics_project/apps/malawi/handlers/abstract/base.py (hunk 1 above)
Comment (gcapalbo) [inline logistics_project/apps/malawi/handlers/abstract/base.py]: updated

File: logistics_project/apps/malawi/handlers/refrigerator_malfunction.py (hunk 2 above)
Comment (gcapalbo) [inline logistics_project/apps/malawi/handlers/refrigerator_malfunction.py]: yeah right now it's responding to the user directly in the [`malfunction_exists`](https://github.com/dimagi/logistics/pull/249/commits/2376dbca6c5d961e97d43dea6e7fe98298cb83cc#diff-dc453c985460c8fbe3dc30a4c0751a72R64) method

File: logistics_project/apps/malawi/models.py (hunk 4 above)
Comment (gcapalbo) [inline logistics_project/apps/malawi/models.py]: i've been avoiding using `choices` for the reasons outlined [here](https://github.com/dimagi/commcare-hq/pull/12286#discussion_r69177555)

File: static/malawi/config.py (hunk 5 above)
Comment (gcapalbo) [inline static/malawi/config.py]: updated

File: static/malawi/config.py (hunk 6 above)
Comment (gcapalbo) [inline static/malawi/config.py]: updated

File: static/malawi/config.py (hunk 7 above)
Comment (gcapalbo) [inline static/malawi/config.py]: updated

File: static/malawi/config.py (hunk 8 above)
Comment (gcapalbo) [inline static/malawi/config.py]: yep, agreed, just tried to define all the messages for now

File: logistics_project/decorators.py [hunk 9]
```diff
@@ -16,3 +18,20 @@ def require_magic_token(request, *args, **kwargs):
             return HttpResponseForbidden("You have to be logged in or have the magic token to do that!")
         return require_magic_token
     return wrapper
+
+
+def require_facility(f):
```
Comment (czue) [inline logistics_project/decorators.py]: nice

File: logistics_project/apps/malawi/extensions/rapidsms/contact.py [hunk 10]
```diff
@@ -44,4 +47,15 @@ def hsa_id(self):
             else:
                 return ""
         else:
-            return ""
\ No newline at end of file
+            return ""
+
+    def translate(self, message):
```
Comment (czue) [inline logistics_project/apps/malawi/extensions/rapidsms/contact.py]: note that I don't *think* this project is used in any language besides english (but I might be wrong about that)

File: logistics_project/apps/malawi/handlers/transfer_epi.py [hunk 11]
```diff
@@ -64,6 +64,10 @@ def handle(self, text):
             if to_facility is None:
                 return
 
+            if from_facility == to_facility:
```
Comment (czue) [inline logistics_project/apps/malawi/handlers/transfer_epi.py]: good call

File: logistics_project/apps/malawi/models.py (hunk 4 above)
Comment (czue) [inline logistics_project/apps/malawi/models.py]: ah, ok. thanks for the heads up.

File: logistics_project/apps/malawi/extensions/rapidsms/contact.py (hunk 10 above)
Comment (czue) [inline logistics_project/apps/malawi/extensions/rapidsms/contact.py]: (i'm about 90% confident)

**Comment by czue** [comment]
looking good. assume rest is mainly tests?

**Comment by gcapalbo** [comment]
yep, tests are all that's left for this

File: logistics_project/apps/malawi/extensions/rapidsms/contact.py (hunk 10 above)
Comment (gcapalbo) [inline logistics_project/apps/malawi/extensions/rapidsms/contact.py]: oh, i do see [this](https://github.com/dimagi/logistics/blob/malawi-dev/logistics_project/locale/sw/LC_MESSAGES/django.po) though?

File: logistics_project/apps/malawi/migrations/0016_auto__add_refrigeratormalfunction.py [hunk 12]
```diff
@@ -247,6 +248,7 @@ def backwards(self, orm):
             'create_date': ('django.db.models.fields.DateTimeField', [], {}),
             'date': ('django.db.models.fields.DateTimeField', [], {}),
             'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
+            'is_facility': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
```
Comment (gcapalbo) [inline logistics_project/apps/malawi/migrations/0016_auto__add_refrigeratormalfunction.py]: (i actually reran the command to recreate the migration but it's good to see this is all that changed)

**Comment by gcapalbo** [comment]
ok, this should be good to go

File: logistics_project/apps/malawi/migrations/0016_auto__add_refrigeratormalfunction.py (hunk 12 above)
Comment (czue) [inline logistics_project/apps/malawi/migrations/0016_auto__add_refrigeratormalfunction.py]: nice

**Comment by czue** [comment]
woop! can offline deploy strategy on this

**Pull request merged**